File: typescript/src/client/validate.ts
/**
 * Offline token validation.
 *
 * Validation runs entirely offline against a pre-configured public-key
 * registry. Order matters and is chosen for minimum-information-leak:
 *
 *   1. Parse + verify signature. Malformed envelopes and bad signatures
 *      never get the chance to be examined for other fields.
 *   2. `nbf` / `exp` with a skew tolerance (default ±60s).
 *   3. Claim shape: required claims present and typed correctly. Without
 *      this, a malicious issuer with a compromised key could hand us a
 *      well-signed but structurally incomplete token.
 *   4. Status: must be `active` or `grace`. Suspended / revoked / expired
 *      get distinct error codes.
 *   5. `force_online_after` — if the deadline is past, validation fails
 *      with `RequiresOnlineRefresh` regardless of signature/exp. The
 *      caller (usually `refresh()`) decides whether to enter grace.
 *   6. Fingerprint match. Checked last so that an expired token produces
 *      `TokenExpired` even when the fingerprint also mismatches (the more
 *      informative error wins).
 */

import type { AlgorithmRegistry, KeyAlgBindings, KeyRecord } from '../crypto/index.ts';
import { decodeUnverified, type LIC1DecodedParts, verify } from '../lic1.ts';

import { clientErrors } from './errors.ts';

export interface ValidateOptions {
  /** Algorithm registry (populated with the same backends the issuer uses).
   *  Clients typically register only the one algorithm they were provisioned
   *  for — e.g. just `ed25519Backend` — to keep the attack surface narrow. */
  readonly registry: AlgorithmRegistry;
  /** Pre-registered kid → alg bindings. Prevents alg-confusion: a token
   *  whose header `alg` disagrees with the bound alg for its kid fails. */
  readonly bindings: KeyAlgBindings;
  /** Public keys the client trusts, keyed by kid. Private key halves MAY
   *  be null — clients never need to sign. */
  readonly keys: ReadonlyMap<string, KeyRecord>;
  /** The device's current fingerprint. Usually derived via
   *  {@link collectFingerprint}; here passed explicitly so validation is
   *  pure (no I/O inside). */
  readonly fingerprint: string;
  /** Current wall-clock unix seconds. Injected for testability — real
   *  callers pass `Math.floor(Date.now() / 1000)`. */
  readonly nowSec: number;
  /** Clock-skew tolerance applied to `nbf` and `exp` checks in both
   *  directions. Default 60s. Set to 0 for strict checks. */
  readonly skewSec?: number;
}

export interface ValidateResult {
  readonly kid: string;
  readonly alg: string;
  readonly license_id: string;
  readonly usage_id: string;
  readonly scope: string;
  readonly status: 'active' | 'grace';
  readonly max_usages: number;
  readonly iat: number;
  readonly nbf: number;
  readonly exp: number;
  /** Absolute unix-seconds deadline, absent when the token doesn't carry one. */
  readonly forceOnlineAfter: number | null;
  readonly entitlements: Readonly<Record<string, unknown>> | null;
}

/**
 * Validate a stored token entirely offline. Returns the decoded result on
 * success; throws a {@link LicensingClientError} with a specific `.code`
 * on any failure.
 */
export async function validate(token: string, opts: ValidateOptions): Promise<ValidateResult> {
  const skew = opts.skewSec ?? 60;

  // 1. Parse + cryptographic verify. `verify` already gates on kid binding
  //    (alg-confusion), unknown kid, and signature validity — we translate
  //    those core errors into client-facing codes below.
  let parts: LIC1DecodedParts;
  try {
    parts = await verify(token, {
      registry: opts.registry,
      bindings: opts.bindings,
      keys: opts.keys,
    });
  } catch (err) {
    throw translateVerifyError(err);
  }

  const { header, payload } = parts;
  const claims = assertClaimShape(payload);

  // 2. nbf / exp.
  if (claims.nbf > opts.nowSec + skew) {
    throw clientErrors.tokenNotYetValid(`nbf=${claims.nbf} now=${opts.nowSec} skew=${skew}`);
  }
  if (claims.exp + skew < opts.nowSec) {
    throw clientErrors.tokenExpired(`exp=${claims.exp} now=${opts.nowSec} skew=${skew}`);
  }

  // 3. Status — usable set is {'active', 'grace'}. Other values would
  //    only appear if the issuer misbehaves, but we handle them
  //    explicitly for a clearer error taxonomy.
  switch (claims.status) {
    case 'active':
    case 'grace':
      break;
    case 'suspended':
      throw clientErrors.licenseSuspended();
    case 'revoked':
      throw clientErrors.licenseRevoked();
    case 'expired':
      throw clientErrors.tokenExpired('token carries status=expired');
    default:
      throw clientErrors.invalidTokenFormat(`unknown status: ${claims.status}`);
  }

  // 4. force_online_after — if present and past, client MUST refresh.
  //    Skew doesn't apply here: the deadline is a hard operational boundary
  //    the issuer chose; treating it fuzzily defeats the purpose.
  if (claims.force_online_after !== null && claims.force_online_after <= opts.nowSec) {
    throw clientErrors.requiresOnlineRefresh(
      `force_online_after=${claims.force_online_after} now=${opts.nowSec}`,
    );
  }

  // 5. Fingerprint match — last so earlier failures dominate. A fingerprint
  //    mismatch usually means the user moved hardware and needs activation,
  //    whereas exp/status failures are recoverable differently.
  if (claims.usage_fingerprint !== opts.fingerprint) {
    throw clientErrors.fingerprintMismatch();
  }

  return {
    kid: header.kid,
    alg: header.alg,
    license_id: claims.license_id,
    usage_id: claims.usage_id,
    scope: claims.scope,
    status: claims.status,
    max_usages: claims.max_usages,
    iat: claims.iat,
    nbf: claims.nbf,
    exp: claims.exp,
    forceOnlineAfter: claims.force_online_after,
    entitlements: claims.entitlements,
  };
}

/**
 * Peek at a token's header + payload without verifying the signature. Useful
 * when the client needs to decide whether a refresh is due ("how much
 * lifetime is left?") before paying the verification cost. Consumers MUST
 * NOT trust the returned data for authorization decisions.
 */
export function peek(token: string): PeekResult {
  try {
    const { header, payload } = decodeUnverified(token);
    const claims = assertClaimShape(payload);
    return {
      kid: header.kid,
      alg: header.alg,
      iat: claims.iat,
      nbf: claims.nbf,
      exp: claims.exp,
      forceOnlineAfter: claims.force_online_after,
    };
  } catch (err) {
    throw clientErrors.invalidTokenFormat(
      `token could not be parsed`,
      err instanceof Error ? err : new Error(String(err)),
    );
  }
}

export interface PeekResult {
  readonly kid: string;
  readonly alg: string;
  readonly iat: number;
  readonly nbf: number;
  readonly exp: number;
  readonly forceOnlineAfter: number | null;
}

// ---------- internals ----------

/** Narrow validated-claims view over the opaque `LIC1Payload`. */
interface RequiredClaims {
  readonly jti: string;
  readonly iat: number;
  readonly nbf: number;
  readonly exp: number;
  readonly scope: string;
  readonly license_id: string;
  readonly usage_id: string;
  readonly usage_fingerprint: string;
  readonly status: string;
  readonly max_usages: number;
  readonly force_online_after: number | null;
  readonly entitlements: Readonly<Record<string, unknown>> | null;
}

function assertClaimShape(payload: Readonly<Record<string, unknown>>): RequiredClaims {
  const req = (key: string) => {
    if (!(key in payload)) {
      throw clientErrors.invalidTokenFormat(`missing required claim: ${key}`);
    }
    return payload[key];
  };
  const asString = (key: string, v: unknown): string => {
    if (typeof v !== 'string') {
      throw clientErrors.invalidTokenFormat(`claim ${key} is not a string`);
    }
    return v;
  };
  const asNumber = (key: string, v: unknown): number => {
    if (typeof v !== 'number' || !Number.isFinite(v)) {
      throw clientErrors.invalidTokenFormat(`claim ${key} is not a finite number`);
    }
    return v;
  };

  return {
    jti: asString('jti', req('jti')),
    iat: asNumber('iat', req('iat')),
    nbf: asNumber('nbf', req('nbf')),
    exp: asNumber('exp', req('exp')),
    scope: asString('scope', req('scope')),
    license_id: asString('license_id', req('license_id')),
    usage_id: asString('usage_id', req('usage_id')),
    usage_fingerprint: asString('usage_fingerprint', req('usage_fingerprint')),
    status: asString('status', req('status')),
    max_usages: asNumber('max_usages', req('max_usages')),
    force_online_after:
      'force_online_after' in payload && payload.force_online_after !== null
        ? asNumber('force_online_after', payload.force_online_after)
        : null,
    entitlements:
      'entitlements' in payload && payload.entitlements !== null
        ? (() => {
            const v = payload.entitlements;
            if (typeof v !== 'object' || Array.isArray(v)) {
              throw clientErrors.invalidTokenFormat('entitlements must be an object');
            }
            return v as Readonly<Record<string, unknown>>;
          })()
        : null,
  };
}

/** Translate a core `verify()` error into the client-side taxonomy. Core
 *  raises `LicensingError` instances with a `.code` matching the issuer-side
 *  enum; we map the subset that can surface on a client offline. */
function translateVerifyError(err: unknown): Error {
  const code = (err as { code?: unknown })?.code;
  const message = err instanceof Error ? err.message : String(err);
  switch (code) {
    case 'UnknownKid':
      return clientErrors.unknownKid(message);
    case 'AlgorithmMismatch':
      return clientErrors.algorithmMismatch('?', message);
    case 'UnsupportedAlgorithm':
      return clientErrors.unsupportedAlgorithm(message);
    case 'TokenSignatureInvalid':
    case 'TokenFormat':
    case 'UnsupportedTokenFormat':
      return clientErrors.invalidTokenFormat(message, err);
    default:
      return clientErrors.invalidTokenFormat(`unexpected verify failure: ${message}`, err);
  }
}

File: licensing/client/validate.go
package client

import (
	"encoding/json"
	"fmt"

	lic "github.com/AnoRebel/licensing/licensing"
)

// ValidateOptions configures offline token validation.
type ValidateOptions struct {
	Registry *lic.AlgorithmRegistry
	Bindings *lic.KeyAlgBindings
	Keys     map[string]lic.KeyRecord
	// JtiLedger, when non-nil, gates each token by jti — a token whose
	// jti was previously recorded surfaces as
	// *ClientError{Code: CodeTokenReplayed}. Meaningful only for online
	// verifiers; offline clients can't share state and should leave
	// this nil. The ledger entry is recorded AFTER all other validate
	// checks pass, so a malformed/expired/wrong-fingerprint token
	// doesn't burn an entry it would have been rejected for anyway.
	JtiLedger JtiLedger
	// ExpectedAudience, when non-empty, pins the audience the token
	// MUST be issued for. The token's `aud` claim — string or array —
	// MUST contain this value. Mismatches surface as
	// CodeAudienceMismatch. Empty means the audience is unpinned and
	// the claim is advisory.
	ExpectedAudience string
	// ExpectedIssuer, when non-empty, pins the issuer the token MUST
	// have been signed by. The token's `iss` claim MUST equal this
	// value. Mismatches surface as CodeIssuerMismatch. Empty means
	// the issuer is unpinned and the claim is advisory.
	ExpectedIssuer string
	Fingerprint    string
	NowSec         int64
	SkewSec        int64 // default 60
}

func (o ValidateOptions) skew() int64 {
	if o.SkewSec > 0 {
		return o.SkewSec
	}
	return 60
}

// ValidateResult carries the decoded and verified claims.
type ValidateResult struct {
	Entitlements     map[string]any
	ForceOnlineAfter *int64
	Iss              *string
	LicenseID        string
	Kid              string
	Alg              lic.KeyAlg
	UsageID          string
	Scope            string
	Status           string
	Aud              []string
	MaxUsages        int64
	Iat              int64
	Nbf              int64
	Exp              int64
}

// PeekResult carries the cheaply-decoded lifetime claims (no sig check).
type PeekResult struct {
	ForceOnlineAfter *int64
	Kid              string
	Alg              lic.KeyAlg
	Iat              int64
	Nbf              int64
	Exp              int64
}

// Validate performs full offline validation of a LIC1 token.
// Check order: (1) parse + sig verify, (2) nbf/exp with skew,
// (3) status in {active, grace}, (4) force_online_after hard deadline,
// (5) fingerprint match.
func Validate(token string, opts ValidateOptions) (*ValidateResult, error) {
	// 1. Parse + signature verify.
	parts, err := lic.Verify(token, lic.VerifyOptions{
		Registry: opts.Registry,
		Bindings: opts.Bindings,
		Keys:     opts.Keys,
	})
	if err != nil {
		return nil, TranslateVerifyError(err)
	}

	// Extract and validate required claims.
	claims, err := assertClaimShape(parts.Payload)
	if err != nil {
		return nil, err
	}

	skew := opts.skew()

	// 2. Not-before / expiry with skew. Uses strict inequality on both
	//    sides per LIC1 spec ("exp < now - skew" → expired; "nbf > now +
	//    skew" → not yet valid) — matches the TS client and keeps the
	//    boundary inclusive of the stated deadline.
	if claims.Nbf > opts.NowSec+skew {
		return nil, TokenNotYetValid(
			fmt.Sprintf("token not valid until %d (now=%d, skew=%d)", claims.Nbf, opts.NowSec, skew))
	}
	if claims.Exp+skew < opts.NowSec {
		return nil, TokenExpired(
			fmt.Sprintf("token expired at %d (now=%d, skew=%d)", claims.Exp, opts.NowSec, skew))
	}

	// 3. Status check.
	switch claims.Status {
	case "active", "grace":
		// OK
	case "revoked":
		return nil, LicenseRevoked("")
	case "suspended":
		return nil, LicenseSuspended("")
	case "expired":
		return nil, TokenExpired("token status is expired")
	default:
		return nil, InvalidTokenFormat(
			fmt.Sprintf("unexpected token status: %s", claims.Status), nil)
	}

	// 4. force_online_after hard deadline (no skew).
	if claims.ForceOnlineAfter != nil && opts.NowSec >= *claims.ForceOnlineAfter {
		return nil, RequiresOnlineRefresh(
			fmt.Sprintf("force_online_after %d has passed (now=%d)",
				*claims.ForceOnlineAfter, opts.NowSec))
	}

	// 5. Fingerprint match.
	if claims.UsageFingerprint != opts.Fingerprint {
		return nil, FingerprintMismatch(
			fmt.Sprintf("token fingerprint %q does not match %q",
				claims.UsageFingerprint, opts.Fingerprint))
	}

	// 6. Audience pin (optional). When the verifier supplies
	//    ExpectedAudience, the token's `aud` MUST contain it. A token
	//    without an `aud` claim fails the pin (the absence is itself
	//    a mismatch in a multi-audience deployment).
	if opts.ExpectedAudience != "" {
		if !audContains(claims.Aud, opts.ExpectedAudience) {
			return nil, AudienceMismatch(
				fmt.Sprintf("expected aud=%s, token has %s",
					opts.ExpectedAudience, audDescription(claims.Aud)))
		}
	}

	// 7. Issuer pin (optional). When the verifier supplies
	//    ExpectedIssuer, the token's `iss` MUST equal it. Tokens
	//    without an `iss` claim fail the pin for the same reason.
	if opts.ExpectedIssuer != "" {
		actual := ""
		if claims.Iss != nil {
			actual = *claims.Iss
		}
		if actual != opts.ExpectedIssuer {
			label := "no iss"
			if claims.Iss != nil {
				label = "iss=" + *claims.Iss
			}
			return nil, IssuerMismatch(
				fmt.Sprintf("expected iss=%s, token has %s",
					opts.ExpectedIssuer, label))
		}
	}

	// 8. Replay-prevention ledger (optional). Records this jti's use
	//    AFTER all other checks have passed so an expired / wrong-
	//    fingerprint / etc. token doesn't burn a ledger entry it would
	//    have been rejected for anyway. Second use of the same jti
	//    surfaces as CodeTokenReplayed.
	if opts.JtiLedger != nil {
		firstUse, err := opts.JtiLedger.RecordJtiUse(claims.Jti, claims.Exp+skew)
		if err != nil {
			return nil, fmt.Errorf("jti ledger: %w", err)
		}
		if !firstUse {
			return nil, TokenReplayed(
				fmt.Sprintf("jti %s was previously recorded by this verifier", claims.Jti))
		}
	}

	return &ValidateResult{
		Kid:              parts.Header.Kid,
		Alg:              parts.Header.Alg,
		LicenseID:        claims.LicenseID,
		UsageID:          claims.UsageID,
		Scope:            claims.Scope,
		Status:           claims.Status,
		MaxUsages:        claims.MaxUsages,
		Iat:              claims.Iat,
		Nbf:              claims.Nbf,
		Exp:              claims.Exp,
		ForceOnlineAfter: claims.ForceOnlineAfter,
		Entitlements:     claims.Entitlements,
		Aud:              claims.Aud,
		Iss:              claims.Iss,
	}, nil
}

// audContains reports whether aud (which may be nil) contains target.
func audContains(aud []string, target string) bool {
	for _, a := range aud {
		if a == target {
			return true
		}
	}
	return false
}

// audDescription is a compact human-readable summary for mismatch
// error messages — avoids emitting a giant audience array into a log.
func audDescription(aud []string) string {
	if aud == nil {
		return "no aud"
	}
	if len(aud) == 1 {
		return "aud=" + aud[0]
	}
	return fmt.Sprintf("aud=[%d values]", len(aud))
}

// Peek performs a synchronous, unverified decode of a LIC1 token. Returns
// the header and lifetime claims. Used by Refresh to cheaply decide if a
// refresh is due without paying the crypto cost.
func Peek(token string) (*PeekResult, error) {
	parts, err := lic.DecodeUnverified(token)
	if err != nil {
		return nil, TranslateVerifyError(err)
	}
	iat, _ := claimInt64(parts.Payload, "iat")
	nbf, _ := claimInt64(parts.Payload, "nbf")
	exp, _ := claimInt64(parts.Payload, "exp")
	foa := claimOptInt64(parts.Payload, "force_online_after")

	return &PeekResult{
		Kid:              parts.Header.Kid,
		Alg:              parts.Header.Alg,
		Iat:              iat,
		Nbf:              nbf,
		Exp:              exp,
		ForceOnlineAfter: foa,
	}, nil
}

// ---------- claim extraction ----------

type requiredClaims struct {
	Entitlements     map[string]any
	ForceOnlineAfter *int64
	Iss              *string
	LicenseID        string
	Jti              string
	Scope            string
	UsageID          string
	UsageFingerprint string
	Status           string
	Aud              []string
	Iat              int64
	Nbf              int64
	Exp              int64
	MaxUsages        int64
}

func assertClaimShape(p lic.LIC1Payload) (*requiredClaims, error) {
	c := &requiredClaims{}
	var ok bool

	c.Jti, ok = claimString(p, "jti")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: jti", nil)
	}
	c.Iat, ok = claimInt64(p, "iat")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: iat", nil)
	}
	c.Nbf, ok = claimInt64(p, "nbf")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: nbf", nil)
	}
	c.Exp, ok = claimInt64(p, "exp")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: exp", nil)
	}
	c.Scope, ok = claimString(p, "scope")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: scope", nil)
	}
	c.LicenseID, ok = claimString(p, "license_id")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: license_id", nil)
	}
	c.UsageID, ok = claimString(p, "usage_id")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: usage_id", nil)
	}
	c.UsageFingerprint, ok = claimString(p, "usage_fingerprint")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: usage_fingerprint", nil)
	}
	c.Status, ok = claimString(p, "status")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: status", nil)
	}
	c.MaxUsages, ok = claimInt64(p, "max_usages")
	if !ok {
		return nil, InvalidTokenFormat("missing or invalid claim: max_usages", nil)
	}

	// Optional claims.
	c.ForceOnlineAfter = claimOptInt64(p, "force_online_after")
	if ents, exists := p["entitlements"]; exists && ents != nil {
		if m, ok := ents.(map[string]any); ok {
			c.Entitlements = m
		}
	}

	// Optional audience: string OR array-of-strings. Wrong types are
	// hard-rejected so a malformed token can't degrade silently to
	// "advisory ignored" mode.
	if aud, exists := p["aud"]; exists && aud != nil {
		switch v := aud.(type) {
		case string:
			c.Aud = []string{v}
		case []any:
			out := make([]string, 0, len(v))
			for _, item := range v {
				s, ok := item.(string)
				if !ok {
					return nil, InvalidTokenFormat("aud array must contain only strings", nil)
				}
				out = append(out, s)
			}
			c.Aud = out
		default:
			return nil, InvalidTokenFormat("aud must be a string or array of strings", nil)
		}
	}

	// Optional issuer: string only.
	if iss, exists := p["iss"]; exists && iss != nil {
		s, ok := iss.(string)
		if !ok {
			return nil, InvalidTokenFormat("iss must be a string", nil)
		}
		c.Iss = &s
	}

	return c, nil
}

func claimString(p lic.LIC1Payload, key string) (string, bool) {
	v, ok := p[key]
	if !ok {
		return "", false
	}
	s, ok := v.(string)
	return s, ok
}

func claimInt64(p lic.LIC1Payload, key string) (int64, bool) {
	v, ok := p[key]
	if !ok {
		return 0, false
	}
	switch x := v.(type) {
	case json.Number:
		i, err := x.Int64()
		return i, err == nil
	case float64:
		return int64(x), x == float64(int64(x))
	case int64:
		return x, true
	case int:
		return int64(x), true
	default:
		return 0, false
	}
}

func claimOptInt64(p lic.LIC1Payload, key string) *int64 {
	v, ok := claimInt64(p, key)
	if !ok {
		return nil
	}
	return &v
}
